lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp

```diff
@@ -0,0 +1,314 @@
+//===------ YieldOnceCheck.cpp - Check usage of yields in accessors  ------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "yield-once-check"
+#include "swift/AST/ASTWalker.h"
+#include "swift/AST/DiagnosticsSIL.h"
+#include "swift/AST/Expr.h"
+#include "swift/AST/Stmt.h"
+#include "swift/SIL/BasicBlockUtils.h"
+#include "swift/SIL/Dominance.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "llvm/ADT/DenseSet.h"
+
+using namespace swift;
+
+namespace {
+
+class YieldOnceCheck : public SILFunctionTransform {
+
+  template <typename... T, typename... U>
+  static InFlightDiagnostic diagnose(ASTContext &Context, SourceLoc loc,
+                                     Diag<T...> diag, U &&... args) {
+    return Context.Diags.diagnose(loc, diag, std::forward<U>(args)...);
+  }
+
+  /// Data-flow analysis state that is associated with basic blocks.
+  struct BBState {
+
+    /// Indicates whether a basic block is encountered before seeing a yield
+    /// (BeforeYield) or after seeing a yield (AfterYield), or in both states
+    /// (Conflict). This enum is a semi-lattice where Conflict is the top and
+    /// the merge of BeforeYield and AfterYield states is Conflict.
+    enum { BeforeYield, AfterYield, Conflict } yieldState = BeforeYield;
```

Comment:
Please give `BBState` a constructor with two arguments, remove the defaulting, and clean up the factories.  It's fine to make it a private constructor, and you may need to add a `getInitialState` factory.

Reply:
Sure. I like that but the only thing that makes default initialization necessary is the DenseMap `bbToStateMap` used in the worklist algorithm. It has the type: `llvm::DenseMap<SILBasicBlock *, BBState>`. I can make the `BBState` optional and remove any default initialization/no-argument constructors. WDYT?

Reply:
I don't think `DenseMap` unconditionally requires its value type to be default-initializable; it's only required if you use methods that need default-initialization, like where you use `try_emplace`.  I could be mistaken, of course.

Reply:
Yes your are right, without those operations, DenseMap doesn't need default constructors. I have pushed a commit to fix these.

Reply:
Looks good, thanks!